### PR TITLE
Fixes #158 Added ISession.BeforeWrite event

### LINKED ISSUE #158: Add ISession.BeforeWrite event
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the CDP4-SDK
- [x] I have searched [open](https://github.com/RHEAGROUP/CDP4-SDK-Community-Edition/issues) and [closed](https://github.com/RHEAGROUP/CDP4-SDK-Community-Edition/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
Sometimes we need some extra actions to be performed just before an `IDal.Write` action occurs in `ISession.Write`.
Please add an event so this requirement can be satisfied.
Make sure that continuation of the `IDal.Write` could be cancelled within an EventHandler

### PR DESCRIPTION
**Pull request by lxatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/CDP4-SDK-Community-Edition/pulls) open
- [x] I have verified that I am following the CDP4-SDK [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/CDP4-SDK-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
Added ISession.BeforeWrite event

### CODE REVIEW

**Review by ahmedatrhea — APPROVED** [review]

File: CDP4Dal/Session.cs [hunk 1]
```diff
@@ -968,4 +981,22 @@ private void AddIterationToOpenList(Guid iterationId, DomainOfExpertise activeDo
             this.AddRdlToOpenList(modelRdl);
         }
     }
+
+    /// <summary>
+    /// Holds <see cref="EventArgs"/> data for a BeforeWrite event.
+    /// </summary>
+    public class BeforeWriteEventArgs : EventArgs
+    {
+        public OperationContainer OperationContainer { get; }
```
Comment (samatrhea) [inline CDP4Dal/Session.cs:990]: please add more xml docs

File: CDP4Dal/Session.cs [hunk 2]
```diff
@@ -968,4 +981,22 @@ private void AddIterationToOpenList(Guid iterationId, DomainOfExpertise activeDo
             this.AddRdlToOpenList(modelRdl);
         }
     }
+
+    /// <summary>
+    /// Holds <see cref="EventArgs"/> data for a BeforeWrite event.
+    /// </summary>
+    public class BeforeWriteEventArgs : EventArgs
```
Comment (samatrhea) [inline CDP4Dal/Session.cs:988]: my preference is to have a file per class

**Review by samatrhea — CHANGES_REQUESTED** [review]

File: CDP4Dal.Tests/Events/BeforeWriteEventArgsTestFixture.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+            Assert.AreEqual(this.operationContainer1, eventArgs1.OperationContainer);
+            Assert.IsNull(eventArgs1.Files);
+
+            var eventArgs2 = new BeforeWriteEventArgs(this.operationContainer2, this.files);
+            Assert.AreEqual(this.operationContainer2, eventArgs2.OperationContainer);
+            Assert.AreEqual(this.files, eventArgs2.Files);
+        }
+    }
+}
```
Comment (samatrhea) [inline CDP4Dal.Tests/Events/BeforeWriteEventArgsTestFixture.cs:68]: new line?

File: CDP4Dal.NetCore.Tests/Events/BeforeWriteEventArgsTestFixture.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+            Assert.AreEqual(this.operationContainer1, eventArgs1.OperationContainer);
+            Assert.IsNull(eventArgs1.Files);
+
+            var eventArgs2 = new BeforeWriteEventArgs(this.operationContainer2, this.files);
+            Assert.AreEqual(this.operationContainer2, eventArgs2.OperationContainer);
+            Assert.AreEqual(this.files, eventArgs2.Files);
+        }
+    }
+}
```
Comment (samatrhea) [inline CDP4Dal.NetCore.Tests/Events/BeforeWriteEventArgsTestFixture.cs:68]: new line

**Review by samatrhea — APPROVED** [review]
just some new lines in some files

**Review by nathanatrhea — APPROVED** [review]

**Review by alexatrhea — APPROVED** [review]

**Pull request merged**